test/base/src/androidTest/kotlin/io/realm/test/shared/QueryTests.kt

```diff
@@ -676,9 +807,131 @@ class QueryTests {
+            // avoid "null vs 0" results when testing
+            cleanUpBetweenProperties()
+        }
+
+        // Iterate over nullable properties containing both null and non-null values - exclude RealmInstant
+        for (nullablePropertyDescriptor in nullablePropertyDescriptorsForSum) {
+            assertions(nullablePropertyDescriptor)
+        }
+    }
+
+    @Test
     fun sum_find() {
-        // TODO
+        val assertions = { propertyDescriptor: PropertyDescriptor ->
+            val expectedSum =
+                expectedSum(propertyDescriptor.clazz, propertyDescriptor.property.returnType)
+            realm.writeBlocking {
+                val sumQuery = query(QuerySample::class)
+                    .sum(propertyDescriptor.property.name, propertyDescriptor.clazz)
+
+                val sumValueBefore = sumQuery.find()
+                if (sumValueBefore is Number) {
+                    assertEquals(0, sumValueBefore.toInt())
+                }
+
+                copyToRealm(getInstance(propertyDescriptor, QuerySample(), 0))
+                copyToRealm(getInstance(propertyDescriptor, QuerySample(), 1))
+
+                val sumValueAfter = sumQuery.find()
+                if (sumValueAfter is Number) {
+                    assertEquals(expectedSum, sumValueAfter)
+                }
+            }
+
+            realm.query(QuerySample::class)
+                .sum(propertyDescriptor.property.name, propertyDescriptor.clazz)
+                .find { sumValue ->
+                    if (sumValue is Number) {
+                        assertEquals(expectedSum, sumValue)
+                    }
+                }
+
+            // Make sure to delete all objects after assertions as aggregators to clean state and
+            // avoid "null vs 0" results when testing
+            cleanUpBetweenProperties()
+        }
+
+        // Iterate over non-nullable properties only - exclude RealmInstant
+        for (propertyDescriptor in propertyDescriptorsForSum) {
+            assertions(propertyDescriptor)
+        }
+
+        // Iterate over nullable properties containing both null and non-null values - exclude RealmInstant
+        for (nullablePropertyDescriptor in nullablePropertyDescriptorsForSum) {
+            assertions(nullablePropertyDescriptor)
+        }
+    }
+
+    @Test
+    fun sum_find_throwsIfRealmInstant() {
+        assertFailsWith<IllegalArgumentException> {
+            realm.query(QuerySample::class)
+                .sum(timestampDescriptor.property.name, timestampDescriptor.clazz)
+                .find() // The sum is only evaluated after obtaining results!
+        }
+
+        assertFailsWith<IllegalArgumentException> {
+            realm.query(QuerySample::class)
+                .sum(nullableTimestampDescriptor.property.name, nullableTimestampDescriptor.clazz)
+                .find() // The sum is only evaluated after obtaining results!
+        }
```

Comment:
Missing some tests here IMO:

* Error when trying to sum a list or object
* What happens on overflow, e.g. If you want a sum as Short but it overflows, should we just overflow or return Short.MAX_VALUE... this needs to be documented in unit tests and documentation.

Reply:
It sounds rather dangerous to automatic truncation based on the type. I would probably prefer to always return the long-value from core and have the user to do the truncation explicitly with `toShort` if they want. This way we get pure Kotlin behavior. Alternatively throw and overflow exception if we do the type conversion and the value doesn't fit.

Reply:
The problem with overflow exists for all types and we cannot only return a long... e.g. there is also double, float and Decimal128 that can be summed.

Personally, I would lean towards respecting the type the user wants, but let it overflow as normal...Throwing seems very heavy-handed and unexpected.

But maybe we should discuss later today?

Reply:
> Personally, I would lean towards respecting the type the user wants, but let it overflow as normal...Throwing seems very heavy-handed and unexpected.

I'm also inclined to take that approach, which is how I implemented it in this PR. The more I think about it the more I believe it's the least complicated one for both us and end-users. All the other options I've tried feel pretty awkward.